mmedit/models/inpaintors/aot_inpaintor.py

```diff
@@ -0,0 +1,279 @@
+        disc (dict): Config for discriminator.
+        loss_gan (dict): Config for adversarial loss.
+        loss_composed_percep (dict): Config for perceptural and style loss with
+            composed image as input.
+        loss_out_percep (dict): Config for perceptural and style loss with
+            direct output as input.
+        loss_l1_valid (dict): Config for l1 loss in the valid region.
+        train_cfg (dict): Configs for training scheduler. `disc_step` must be
+            contained for indicates the discriminator updating steps in each
+            training step.
+        test_cfg (dict): Configs for testing scheduler.
+        pretrained (str): Path for pretrained model. Default None.
+    """
+
+    def __init__(self,
+                 encdec,
+                 disc=None,
+                 loss_gan=None,
+                 loss_composed_percep=None,
+                 loss_out_percep=False,
+                 loss_l1_valid=None,
+                 train_cfg=None,
+                 test_cfg=None,
+                 pretrained=None):
+        super().__init__(
+            encdec,
+            disc=disc,
+            loss_gan=loss_gan,
+            loss_composed_percep=loss_composed_percep,
+            loss_out_percep=loss_out_percep,
+            loss_l1_valid=loss_l1_valid,
+            train_cfg=train_cfg,
+            test_cfg=test_cfg,
+            pretrained=pretrained)
+
+    def forward_train_d(self, data_batch, is_real, is_disc, **kwargs):
+        """Forward function in discriminator training step.
+        In this function, we compute the prediction for each data batch (real
+        or fake). Meanwhile, the standard gan loss will be computed with
+        several proposed losses fro stable training.
+        Args:
+            data (torch.Tensor): Batch of real data or fake data.
+            is_real (bool): If True, the gan loss will regard this batch as
+                real data. Otherwise, the gan loss will regard this batch as
+                fake data.
+            is_disc (bool): If True, this function is called in discriminator
+                training step. Otherwise, this function is called in generator
+                training step. This will help us to compute different types of
+                adversarial loss, like LSGAN.
+            **kwargs (keyword arguments).
+        Returns:
+            dict: Contains the loss items computed in this function.
+        """
+
+        pred = self.disc(data_batch)
+        loss_ = self.loss_gan(pred, is_real, is_disc, **kwargs)
+
+        loss = dict(real_loss=loss_) if is_real else dict(fake_loss=loss_)
+
+        if self.with_disc_shift_loss:
+            loss_d_shift = self.loss_disc_shift(loss_)
+            # 0.5 for average the fake and real data
+            loss.update(loss_disc_shift=loss_d_shift * 0.5)
+
+        return loss
+
+    def generator_loss(self, fake_res, fake_img, data_batch):
```

Comment:
This should be identical to that in `OneStageInpaintor`?

Reply:
I overwrote the generator_loss function in aot inpaintor because there are a few functions within it which take different arguments from the parent onestageinpaintor. 

For example, in aot inpaintor the l1_valid_loss is not weighted by the mask, so weight is not passed into the function self.loss_l1_valid:

![image](https://user-images.githubusercontent.com/33197366/149704439-3059f8e5-020d-4e2f-9e7b-f74a16c1dabb.png)

whereas in onestageinpaintor the l1 valid loss is weighted by the mask:

![image](https://user-images.githubusercontent.com/33197366/149704496-e60a1bc8-7956-436b-9899-11efb1e8b95d.png)

Reply:
Oh okay, I missed that. Thanks for your clarification.